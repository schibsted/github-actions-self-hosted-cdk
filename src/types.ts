import { StackProps } from 'aws-cdk-lib';

export interface WebhookEnvironment {
  templateId: string;
  templateVersion: string;
  context: string;
  subnetId: string;
  spot: string;
  webhookSecretSsmPath: string;
  webhookSecretSsmArn: string;
}

export interface Context {
  name: string;
  scope: string;
  group?: string;
  timeout?: string;
  spot?: boolean;
  webhookSecretSsmPath: string;
}

export interface GithubActionsRunnersProps extends StackProps {
  tokenSsmPath: string;
  runnerVersion?: string;
  debugMode?: boolean;
  private?: boolean;
  contexts?: Context[];
}
