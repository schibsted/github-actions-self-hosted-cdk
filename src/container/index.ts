import { Stack } from 'aws-cdk-lib';
import { IVpc } from 'aws-cdk-lib/aws-ec2';
import {
  Cluster,
  FargateTaskDefinition,
  ContainerImage,
  LogDrivers,
  Secret,
} from 'aws-cdk-lib/aws-ecs';
import { StringParameter } from 'aws-cdk-lib/aws-ssm';
import path from 'path';
import { GithubActionsRunnerParams } from '../types';

export const setupContainerRunners = (
  stack: Stack,
  props: GithubActionsRunnerParams,
  vpc: IVpc,
) => {
  const cluster = new Cluster(stack, 'GitHubActionsRunnerCluster', {
    vpc,
    enableFargateCapacityProviders: true,
    clusterName: 'GitHubActionsRunnerCluster',
  });

  const taskDefinition = new FargateTaskDefinition(stack, 'TaskDef', {});
  taskDefinition.addContainer('Container', {
    containerName: 'Runner',
    image: ContainerImage.fromAsset(path.resolve(__dirname, '.')),
    logging: LogDrivers.awsLogs({ streamPrefix: 'GitHubActionsRunner' }),
    environment: {
      RUNNER_CONTEXT: props.context,
      RUNNER_TIMEOUT: props.runnerTimeout ?? '60m',
    },
    secrets: {
      GITHUB_TOKEN: Secret.fromSsmParameter(
        StringParameter.fromSecureStringParameterAttributes(
          stack,
          'GitHubAccessToken',
          {
            parameterName: props.tokenSsmPath,
            version: 0,
          },
        ),
      ),
    },
  });

  return {
    clusterName: cluster.clusterName,
    clusterArn: cluster.clusterArn,
    taskDefinition: taskDefinition.family,
  };
};
