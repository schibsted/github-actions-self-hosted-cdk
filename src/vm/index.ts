import { Stack } from 'aws-cdk-lib';
import {
  InstanceInitiatedShutdownBehavior,
  InstanceType,
  UserData,
  MachineImage,
  LaunchTemplate,
  ISecurityGroup,
} from 'aws-cdk-lib/aws-ec2';
import {
  PolicyStatement,
  Role,
  ServicePrincipal,
  PolicyDocument,
  CfnInstanceProfile,
  ManagedPolicy,
} from 'aws-cdk-lib/aws-iam';
import {
  CfnImage,
  CfnImageRecipe,
  CfnInfrastructureConfiguration,
  CfnComponent,
} from 'aws-cdk-lib/aws-imagebuilder';
import { readFileSync } from 'fs';
import path from 'path';
import { GithubActionsRunnersProps, Context } from '../types';

export const buildImage = (stack: Stack, props: GithubActionsRunnersProps) => {
  const region = props.env?.region!;

  const component = new CfnComponent(stack, 'GithubActionsRunnerComponent', {
    name: stack.artifactId,
    version: '1.0.2',
    platform: 'Linux',
    data: readFileSync(path.resolve(__dirname, './component.yml'), 'utf8'),
  });

  const recipe = new CfnImageRecipe(stack, 'Recipe', {
    name: stack.artifactId,
    version: '0.0.4',
    parentImage: `arn:aws:imagebuilder:${region}:aws:image/ubuntu-server-20-lts-x86/x.x.x`,
    components: [
      {
        componentArn: `arn:aws:imagebuilder:${region}:aws:component/update-linux/1.0.0`,
      },
      {
        componentArn: `arn:aws:imagebuilder:${region}:aws:component/docker-ce-ubuntu/1.0.0`,
      },
      {
        componentArn: component.attrArn,
        parameters: [
          {
            name: 'RunnerVersion',
            value: [props.runnerVersion ?? '2.286.0'],
          },
        ],
      },
    ],
  });

  const instanceProfile = new CfnInstanceProfile(stack, 'AmiInstanceProfile', {
    path: '/executionServiceEC2Role/',
    instanceProfileName: stack.artifactId,
    roles: [
      new Role(stack, 'InstanceProfileRole', {
        assumedBy: new ServicePrincipal('ec2.amazonaws.com'),
        managedPolicies: [
          ManagedPolicy.fromAwsManagedPolicyName(
            'AmazonSSMManagedInstanceCore',
          ),
          ManagedPolicy.fromAwsManagedPolicyName(
            'EC2InstanceProfileForImageBuilder',
          ),
        ],
      }).roleName,
    ],
  });

  const infraConfig = new CfnInfrastructureConfiguration(
    stack,
    'AmiInfraConfig',
    {
      name: stack.artifactId,
      instanceProfileName: instanceProfile.instanceProfileName!,
    },
  );
  infraConfig.addDependsOn(instanceProfile);

  const ami = new CfnImage(stack, 'Ami', {
    imageRecipeArn: recipe.attrArn,
    infrastructureConfigurationArn: infraConfig.attrArn,
  });

  return {
    imageId: ami.attrImageId,
  };
};

export const setupRunners = (
  stack: Stack,
  props: GithubActionsRunnersProps,
  context: Context,
  securityGroup: ISecurityGroup,
  imageId: string,
) => {
  const region = props.env?.region!;
  const userDataScript = readFileSync(
    path.resolve(__dirname, './entrypoint.sh'),
    'utf8',
  )
    .replace('$AWS_REGION', region)
    .replace('$GITHUB_TOKEN_SSM_PATH', context.tokenSsmPath)
    .replace('$GITHUB_HOST', props.githubHost)
    .replace('$RUNNER_SCOPE', context.scope)
    .replace('$RUNNER_GROUP', context.group ?? 'default')
    .replace('$RUNNER_TIMEOUT', context.timeout ?? '60m');
  const template = new LaunchTemplate(stack, `LaunchTemplate/${context.name}`, {
    launchTemplateName: `${stack.artifactId}/${context.name}`,
    userData: UserData.custom(userDataScript),
    instanceType: new InstanceType('t3.micro'),
    machineImage: MachineImage.genericLinux({
      [region]: imageId,
    }),
    instanceInitiatedShutdownBehavior:
      InstanceInitiatedShutdownBehavior.TERMINATE,
    securityGroup,
    role: new Role(stack, `RunnerRole${context.name}`, {
      assumedBy: new ServicePrincipal('ec2.amazonaws.com'),
      inlinePolicies: {
        'ssm-policy': new PolicyDocument({
          statements: [
            new PolicyStatement({
              actions: ['ssm:*'],
              resources: ['*'],
            }),
          ],
        }),
      },
    }),
  });

  return {
    templateId: template.launchTemplateId ?? '',
    templateVersion: template.latestVersionNumber,
  };
};
