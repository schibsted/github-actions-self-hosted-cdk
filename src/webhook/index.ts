import { Stack } from 'aws-cdk-lib';
import { RestApi, LambdaIntegration } from 'aws-cdk-lib/aws-apigateway';
import { Function, Runtime, Code } from 'aws-cdk-lib/aws-lambda';
import { PolicyStatement, Policy } from 'aws-cdk-lib/aws-iam';
import path from 'path';
import { WebhookEnvironment } from '../types';

export const setupWekhook = (stack: Stack, env: WebhookEnvironment) => {
  const func = new Function(stack, 'WebhookLambda', {
    runtime: Runtime.NODEJS_14_X,
    handler: 'handler.handler',
    code: Code.fromAsset(path.join(__dirname, '.')),
    environment: {
      ...env,
    },
  });
  func.role?.attachInlinePolicy(
    new Policy(stack, 'RunTaskPolicy', {
      statements: [
        new PolicyStatement({
          actions: ['ecs:RunTask'],
          resources: ['*'],
          conditions: {
            ArnEquals: {
              'ecs:cluster': env.clusterArn,
            },
          },
        }),
        new PolicyStatement({
          actions: ['ec2:RunInstances', 'ec2:CreateTags'],
          resources: ['*'],
        }),
        new PolicyStatement({
          actions: ['iam:PassRole'],
          resources: ['*'],
          conditions: {
            StringLike: {
              'iam:PassedToService': 'ecs-tasks.amazonaws.com',
            },
          },
        }),
        new PolicyStatement({
          actions: ['iam:PassRole'],
          resources: ['*'],
          conditions: {
            StringLike: {
              'iam:PassedToService': 'ec2.amazonaws.com',
            },
          },
        }),
      ],
    }),
  );
  const api = new RestApi(stack, 'api');
  const resource = api.root.addResource('webhook');
  resource.addMethod('POST', new LambdaIntegration(func));
};
