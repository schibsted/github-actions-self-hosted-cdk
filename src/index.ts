import { Stack } from 'aws-cdk-lib';
import { Vpc, Peer, Port, SecurityGroup } from 'aws-cdk-lib/aws-ec2';
import { Construct } from 'constructs';
import { GithubActionsRunnerParams } from './types';
import { setupVMRunners } from './vm';
import { setupContainerRunners } from './container';
import { setupWekhook } from './webhook';

export class GithubActionsRunnerStack extends Stack {
  constructor(scope: Construct, id: string, props: GithubActionsRunnerParams) {
    super(scope, id, props);

    const vpc = Vpc.fromLookup(this, 'Vpc', {
      isDefault: true,
    });
    const securityGroup = new SecurityGroup(this, 'SecurityGroup', {
      vpc,
      allowAllOutbound: true,
      securityGroupName: 'gh-actions-runner-sg',
    });
    securityGroup.addIngressRule(Peer.anyIpv4(), Port.tcp(22));

    const vm = setupVMRunners(this, props, securityGroup);
    const container = setupContainerRunners(this, props, vpc);
    setupWekhook(this, {
      ...vm,
      ...container,
      securityGroup: securityGroup.securityGroupId,
      subnets: vpc.publicSubnets.map(x => x.subnetId).join(','),
    });
  }
}
