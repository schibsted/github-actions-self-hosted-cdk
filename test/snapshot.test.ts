import { Stack } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { GithubActionsRunnerStack } from '../src/index';

test('Snapshot test', () => {
  const stack = new GithubActionsRunnerStack(new Stack(), 'TestStack', {
    env: {
      account: 'abc-123',
      region: 'eu-north-1',
    },
    context: 'spt-mediaplatform-labs',
    tokenSsmPath: '/github/actions/token',
  });
  expect(Template.fromStack(stack).toJSON()).toMatchSnapshot();
});
