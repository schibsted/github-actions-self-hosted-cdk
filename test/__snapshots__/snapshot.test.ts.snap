// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot test 1`] = `
Object {
  "Outputs": Object {
    "apiEndpoint9349E63C": Object {
      "Value": Object {
        "Fn::Join": Array [
          "",
          Array [
            "https://",
            Object {
              "Ref": "apiC8550315",
            },
            ".execute-api.eu-north-1.",
            Object {
              "Ref": "AWS::URLSuffix",
            },
            "/",
            Object {
              "Ref": "apiDeploymentStageprod896C8101",
            },
            "/",
          ],
        ],
      },
    },
  },
  "Parameters": Object {
    "BootstrapVersion": Object {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": Object {
    "Ami": Object {
      "Properties": Object {
        "ImageRecipeArn": Object {
          "Fn::GetAtt": Array [
            "Recipe",
            "Arn",
          ],
        },
        "InfrastructureConfigurationArn": Object {
          "Fn::GetAtt": Array [
            "AmiInfraConfig",
            "Arn",
          ],
        },
      },
      "Type": "AWS::ImageBuilder::Image",
    },
    "AmiInfraConfig": Object {
      "DependsOn": Array [
        "AmiInstanceProfile",
      ],
      "Properties": Object {
        "InstanceProfileName": "GithubActionsRunnerInstanceProfile",
        "Name": "GithubActionsRunnerAmiInfraConfig",
      },
      "Type": "AWS::ImageBuilder::InfrastructureConfiguration",
    },
    "AmiInstanceProfile": Object {
      "Properties": Object {
        "InstanceProfileName": "GithubActionsRunnerInstanceProfile",
        "Path": "/executionServiceEC2Role/",
        "Roles": Array [
          Object {
            "Ref": "InstanceProfileRole5BFF5D1E",
          },
        ],
      },
      "Type": "AWS::IAM::InstanceProfile",
    },
    "GithubActionsRunnerComponent": Object {
      "Properties": Object {
        "Data": "name: InstallRunner
description: Install Github Actions runner
schemaVersion: 1.0
parameters:
  - RunnerVersion:
      type: string
      default: '2.286.0'
      description: Github Actions Runner version
phases:
  - name: build
    steps:
      - name: InstallRunner
        action: ExecuteBash
        inputs:
          commands:
            - sudo apt-get update -y
            - sudo apt-get install -y curl jq awscli
            - cd /home/ubuntu
            - sudo -u ubuntu curl -sSLO https://github.com/actions/runner/releases/download/v{{ RunnerVersion }}/actions-runner-linux-x64-{{ RunnerVersion }}.tar.gz
            - sudo -u ubuntu tar -zxvf actions-runner-linux-x64-{{ RunnerVersion }}.tar.gz
            - sudo -u ubuntu rm -f actions-runner-linux-x64-{{ RunnerVersion }}.tar.gz
            - sudo ./bin/installdependencies.sh",
        "Name": "Install Runner",
        "Platform": "Linux",
        "Version": "1.0.2",
      },
      "Type": "AWS::ImageBuilder::Component",
    },
    "InstanceProfileRole5BFF5D1E": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "ec2.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/AmazonSSMManagedInstanceCore",
              ],
            ],
          },
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/EC2InstanceProfileForImageBuilder",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "LaunchGithubActionsRunner65ABA738": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "ec2:RunInstances",
                "ec2:CreateTags",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
            Object {
              "Action": "iam:PassRole",
              "Condition": Object {
                "StringLike": Object {
                  "iam:PassedToService": "ec2.amazonaws.com",
                },
              },
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "LaunchGithubActionsRunner65ABA738",
        "Roles": Array [
          Object {
            "Ref": "WebhookLambdaServiceRoleDB3A5765",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "LaunchTemplate04EC5460": Object {
      "Properties": Object {
        "LaunchTemplateData": Object {
          "IamInstanceProfile": Object {
            "Arn": Object {
              "Fn::GetAtt": Array [
                "LaunchTemplateProfile94AA77CE",
                "Arn",
              ],
            },
          },
          "ImageId": Object {
            "Fn::GetAtt": Array [
              "Ami",
              "ImageId",
            ],
          },
          "InstanceInitiatedShutdownBehavior": "terminate",
          "InstanceType": "t3.micro",
          "SecurityGroupIds": Array [
            Object {
              "Fn::GetAtt": Array [
                "SecurityGroupDD263621",
                "GroupId",
              ],
            },
          ],
          "TagSpecifications": Array [
            Object {
              "ResourceType": "instance",
              "Tags": Array [
                Object {
                  "Key": "Name",
                  "Value": "Default/TestStack/LaunchTemplate",
                },
              ],
            },
            Object {
              "ResourceType": "volume",
              "Tags": Array [
                Object {
                  "Key": "Name",
                  "Value": "Default/TestStack/LaunchTemplate",
                },
              ],
            },
          ],
          "UserData": Object {
            "Fn::Base64": "#!/bin/bash

AWS_REGION=eu-north-1
AWS_INSTANCE_TYPE=$(curl http://169.254.169.254/latest/meta-data/instance-type)
GH_TOKEN_SSM_PATH=/github/actions/token
RUNNER_CONTEXT=spt-mediaplatform-labs
RUNNER_TIMEOUT=60m
RUNNER_GROUP=default

RUNNER_WORKDIR=_work

cd /home/ubuntu

deregister_runner() {
  echo \\"Exit signal caught, deregistering runner...\\"
  sudo -u ubuntu ./config.sh remove --unattended --token \\"\${RUNNER_TOKEN}\\"
}

init() {
  GITHUB_TOKEN=$(aws ssm get-parameters --name \${GH_TOKEN_SSM_PATH} --region \${AWS_REGION} --with-decryption | jq -r .Parameters[0].Value)
  AUTH_HEADER=\\"Authorization: token \${GITHUB_TOKEN}\\"
  ORG=$(cut -d/ -f1 <<< \${RUNNER_CONTEXT})
  REPO=$(cut -d/ -f2 <<< \${RUNNER_CONTEXT})

  if [[ -z \\"\${REPOSITORY}\\" ]]; then
    TOKEN_REGISTRATION_URL=\\"https://github.schibsted.io/api/v3/orgs/\${ORG}/actions/runners/registration-token\\"
  else
    TOKEN_REGISTRATION_URL=\\"https://github.schibsted.io/api/v3/repos/\${ORG}/\${REPO}/actions/runners/registration-token\\"
  fi

  RUNNER_TOKEN=\\"$(sudo -u ubuntu curl -XPOST -fsSL \\\\
    -H \\"Accept: application/vnd.github.v3+json\\" \\\\
    -H \\"\${AUTH_HEADER}\\" \\\\
    \\"\${TOKEN_REGISTRATION_URL}\\" \\\\
    | jq -r '.token')\\"
}

configure() {
  sudo -u ubuntu ./config.sh --url \\"https://github.schibsted.io/\${RUNNER_CONTEXT}\\" \\\\
    --token \\"\${RUNNER_TOKEN}\\" \\\\
    --labels \\"vm:\${AWS_INSTANCE_TYPE}\\" \\\\
    --work \\"\${RUNNER_WORKDIR}\\" \\\\
    --runnergroup \\"\${RUNNER_GROUP}\\" \\\\
    --unattended \\\\
    --ephemeral
}

run() {
  trap deregister_runner SIGINT SIGQUIT SIGTERM INT QUIT TERM
  sudo -u ubuntu timeout --signal=15 \\"\${RUNNER_TIMEOUT}\\" ./run.sh & wait $! && return
  deregister_runner
}

echo \\"Init...\\"
init

echo \\"Configuring...\\"
configure

echo \\"Running...\\"
run

echo \\"Shutting down...\\"
sudo shutdown -h now",
          },
        },
        "LaunchTemplateName": "GithubActionsRunnerTemplate",
      },
      "Type": "AWS::EC2::LaunchTemplate",
    },
    "LaunchTemplateProfile94AA77CE": Object {
      "Properties": Object {
        "Roles": Array [
          Object {
            "Ref": "RunnerRoleBA0605C0",
          },
        ],
      },
      "Type": "AWS::IAM::InstanceProfile",
    },
    "Recipe": Object {
      "Properties": Object {
        "Components": Array [
          Object {
            "ComponentArn": "arn:aws:imagebuilder:eu-north-1:aws:component/update-linux/1.0.0",
          },
          Object {
            "ComponentArn": "arn:aws:imagebuilder:eu-north-1:aws:component/docker-ce-ubuntu/1.0.0",
          },
          Object {
            "ComponentArn": Object {
              "Fn::GetAtt": Array [
                "GithubActionsRunnerComponent",
                "Arn",
              ],
            },
            "Parameters": Array [
              Object {
                "Name": "RunnerVersion",
                "Value": Array [
                  "2.286.0",
                ],
              },
            ],
          },
        ],
        "Name": "GithubActionsRunnerAmiRecipe",
        "ParentImage": "arn:aws:imagebuilder:eu-north-1:aws:image/ubuntu-server-20-lts-x86/x.x.x",
        "Version": "0.0.4",
      },
      "Type": "AWS::ImageBuilder::ImageRecipe",
    },
    "RunnerRoleBA0605C0": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "ec2.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Policies": Array [
          Object {
            "PolicyDocument": Object {
              "Statement": Array [
                Object {
                  "Action": "ssm:*",
                  "Effect": "Allow",
                  "Resource": "*",
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "ssm-policy",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "SecurityGroupDD263621": Object {
      "Properties": Object {
        "GroupDescription": "Default/TestStack/SecurityGroup",
        "GroupName": "github-actions-runners",
        "SecurityGroupEgress": Array [
          Object {
            "CidrIp": "0.0.0.0/0",
            "Description": "Allow all outbound traffic by default",
            "IpProtocol": "-1",
          },
        ],
        "SecurityGroupIngress": Array [
          Object {
            "CidrIp": "0.0.0.0/0",
            "Description": "from 0.0.0.0/0:22",
            "FromPort": 22,
            "IpProtocol": "tcp",
            "ToPort": 22,
          },
        ],
        "VpcId": "vpc-12345",
      },
      "Type": "AWS::EC2::SecurityGroup",
    },
    "WebhookLambda118EFE5D": Object {
      "DependsOn": Array [
        "WebhookLambdaServiceRoleDB3A5765",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": "cdk-hnb659fds-assets-abc-123-eu-north-1",
          "S3Key": "b4495ced0aa18646ab37f42d23c98fd045de74b3bcc8e8e764a7579d5f4fc9f5.zip",
        },
        "Environment": Object {
          "Variables": Object {
            "context": "spt-mediaplatform-labs",
            "templateId": Object {
              "Ref": "LaunchTemplate04EC5460",
            },
            "templateVersion": Object {
              "Fn::GetAtt": Array [
                "LaunchTemplate04EC5460",
                "LatestVersionNumber",
              ],
            },
          },
        },
        "Handler": "handler.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "WebhookLambdaServiceRoleDB3A5765",
            "Arn",
          ],
        },
        "Runtime": "nodejs14.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "WebhookLambdaServiceRoleDB3A5765": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "apiAccount57E28B43": Object {
      "DependsOn": Array [
        "apiC8550315",
      ],
      "Properties": Object {
        "CloudWatchRoleArn": Object {
          "Fn::GetAtt": Array [
            "apiCloudWatchRoleAC81D93E",
            "Arn",
          ],
        },
      },
      "Type": "AWS::ApiGateway::Account",
    },
    "apiC8550315": Object {
      "Properties": Object {
        "Name": "api",
      },
      "Type": "AWS::ApiGateway::RestApi",
    },
    "apiCloudWatchRoleAC81D93E": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "apigateway.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AmazonAPIGatewayPushToCloudWatchLogs",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "apiDeployment149F129491d0c069ff1ae77e5ca543966e2abdea": Object {
      "DependsOn": Array [
        "apiwebhookPOSTB12B11A0",
        "apiwebhook7D40138A",
      ],
      "Properties": Object {
        "Description": "Automatically created by the RestApi construct",
        "RestApiId": Object {
          "Ref": "apiC8550315",
        },
      },
      "Type": "AWS::ApiGateway::Deployment",
    },
    "apiDeploymentStageprod896C8101": Object {
      "DependsOn": Array [
        "apiAccount57E28B43",
      ],
      "Properties": Object {
        "DeploymentId": Object {
          "Ref": "apiDeployment149F129491d0c069ff1ae77e5ca543966e2abdea",
        },
        "RestApiId": Object {
          "Ref": "apiC8550315",
        },
        "StageName": "prod",
      },
      "Type": "AWS::ApiGateway::Stage",
    },
    "apiwebhook7D40138A": Object {
      "Properties": Object {
        "ParentId": Object {
          "Fn::GetAtt": Array [
            "apiC8550315",
            "RootResourceId",
          ],
        },
        "PathPart": "webhook",
        "RestApiId": Object {
          "Ref": "apiC8550315",
        },
      },
      "Type": "AWS::ApiGateway::Resource",
    },
    "apiwebhookPOSTApiPermissionTestStackapiF0E8311DPOSTwebhook95871F4F": Object {
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Fn::GetAtt": Array [
            "WebhookLambda118EFE5D",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": Object {
          "Fn::Join": Array [
            "",
            Array [
              "arn:",
              Object {
                "Ref": "AWS::Partition",
              },
              ":execute-api:eu-north-1:abc-123:",
              Object {
                "Ref": "apiC8550315",
              },
              "/",
              Object {
                "Ref": "apiDeploymentStageprod896C8101",
              },
              "/POST/webhook",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "apiwebhookPOSTApiPermissionTestTestStackapiF0E8311DPOSTwebhookB4A914DC": Object {
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Fn::GetAtt": Array [
            "WebhookLambda118EFE5D",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": Object {
          "Fn::Join": Array [
            "",
            Array [
              "arn:",
              Object {
                "Ref": "AWS::Partition",
              },
              ":execute-api:eu-north-1:abc-123:",
              Object {
                "Ref": "apiC8550315",
              },
              "/test-invoke-stage/POST/webhook",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "apiwebhookPOSTB12B11A0": Object {
      "Properties": Object {
        "AuthorizationType": "NONE",
        "HttpMethod": "POST",
        "Integration": Object {
          "IntegrationHttpMethod": "POST",
          "Type": "AWS_PROXY",
          "Uri": Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":apigateway:eu-north-1:lambda:path/2015-03-31/functions/",
                Object {
                  "Fn::GetAtt": Array [
                    "WebhookLambda118EFE5D",
                    "Arn",
                  ],
                },
                "/invocations",
              ],
            ],
          },
        },
        "ResourceId": Object {
          "Ref": "apiwebhook7D40138A",
        },
        "RestApiId": Object {
          "Ref": "apiC8550315",
        },
      },
      "Type": "AWS::ApiGateway::Method",
    },
  },
  "Rules": Object {
    "CheckBootstrapVersion": Object {
      "Assertions": Array [
        Object {
          "Assert": Object {
            "Fn::Not": Array [
              Object {
                "Fn::Contains": Array [
                  Array [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  Object {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
